docs/design/features/ScalableApproximateCounting.md

```diff
@@ -0,0 +1,185 @@
+
+### Accuracy
+
+As noted, *interlocked* is perfectly accurate. If we compare *racing* and *scalable* to interlocked we see results similar to the following:
+
+![Counter Accuracy](counter-accuracy.png)
+
+Here the X axis is the intended final value of the count, where counting is done by #processors (here 12) threads each doing nothing more than incrementing the counter, and the Y axis shows the mean counter value along with max and min values, scaled by the total expected count.
+
+As expected, *Interlocked* is just a horizontal line at Y=1.0.
+
+*Racing* is losing counts even with a small final count of 12, and the loss percentage gets worse and worse the higher we try to count, ending up with at best 20% of the final count value.
+
+*Scalable* is perfectly accurate up to 8192, then starts to deviate somewhat, but the total relative deviation remains constrained.
+
+If we remove *Racing* from the picture and zoom in, we see that the error is indeed 2% for the 5/95 spread, and 3% for worst case, and that on average the count is very accurate.
+
+![Counter Accuracy Detail](counter-accuracy-detail.png)
+
+We can also adjust the accuracy by changing when we switch to probabilistic counting; the higher the switchover point, the more accurate the results are overall. In the below we vary the switchover point from $2^{10}$ to $2^{15}$ and plot the 5th and 95th percentiles relative to the accurate count:
+
+![Alt text](counter-accuracy-tunable.png)
+
+### Scalability
+
+Benchmarking the same set of computations with, we end up with the following (here the switchover is at 8192):
+
+![Counter Scaling](counter-scaling.png)
+
+So *scalable* and *racing* have similar costs, while *interlocked* becomes quite a bit more costly at high counts.
+
+### Notes
+
+* Quite likely the benchmark has too much fixed overhead and could be improved, and if so we might see distinctions even for lower total counts.
+* It was critical for *scalable* to use per-thread RNGs and to seed them each differently.
+
+Mathematically we can model the *scalable* update distribution as a binomial distribution. If we make $N$ counter updates with probability $P$, the expected number of updates is $NP$ and the standard deviation in the number of updates is $\sqrt{NP(1-P)}$.
+
+So if we start probabilistically incrementing by $2$ with probability $1/2$ at $8192$, then after $8192$ probabilistic updates we have added an expected value of $8192 \cdot 2 \cdot 1/2 = 8192$ to the counter.
+
+The variance in the actual number of updates is $\sqrt{2^{13} \cdot 1/2 \cdot (1-1/2)} = \sqrt{2^{11}} \approx 45$. Each update is by 2, so the two standard deviation expected range for the change in the counter value is $2 \cdot 2 \cdot 45 \approx 180$. The relative error range is thus $\pm 180 / 8192 \approx \pm 0.022$. This is in reasonable agreement with the empirical study above.
```

Comment:
I think there's an interesting thing (coincidence?) going on here.  The empirical study is going to include the effects of perfect measurement from 0-8192, so the reported empirical value for 16k is going to be half this relative error, or 0.011.  This matches the graph very well.

This calculation for each additional section quickly goes to 0.03-0.031.  However, again, the measured cumulative error is going to be smaller.  It turns out to be close to that initial calculation.

One thing I can't figure out is the graph where you vary the starting point.  If I do the calculation for 10, I get 0.0625 (and even halving it for the cumulative effect I get ~0.03).  However, the first data point off the center line is 1.005.  (Or maybe I should be looking at the next one, which is close to 1.03?  The x-axis seems confusing here because that "first" point appears to be over 1k when it should be over 2k since 1k is the crossover point.  Maybe?  And I guess that I need to point out that the non-powers-of-2 on the x-axis are a little worrying :))

Reply:
Ah, good catch. I think there's a bug in my simulation code, the log should be
```C#
   int logCount = 32 - (int) uint.LeadingZeroCount(count);
```
So the simulation starts probabilistic mode sooner than it should. Let me rerun this and see if the graph makes more sense.

Reply:
No, that's not the issue, the threshold is OK. When I added a parallel mode to the simulation to I wanted to make sure there was evenly divisible counting I rounded up the count to the next multiple of the number of CPUs. So that skews the data a bit as you noted, and so yes you should look at "the next one".

Here is a revised plot without the extra counts:

![image](https://user-images.githubusercontent.com/10121823/230691797-cd46efa1-1e9b-4b76-baf0-d7ab606208f6.png)

Reply:
That chart matches what I found. It's fascinating how such a seemingly simple idea works out so well.